### PR TITLE
Posix standard implementation for pse51

### PR DESCRIPTION
**Pull request by xiangxistu**
## 拉取/合并请求描述：(PR description)

[
  * add some posix's interfaces that we haven't before.
  * these PR have passed the interface definition test across gcc platfrom;
  * have tested base on qemu-a9 and stm32h750-art-pi.

]

以下的内容不应该在提交PR时的message修改，修改下述message，PR会被直接关闭。请在提交PR后，浏览器查看PR并对以下检查项逐项check，没问题后逐条在页面上打钩。
The following content must not be changed in the submitted PR message. Otherwise, the PR will be closed immediately. After submitted PR, please use a web browser to visit PR, and check items one by one, and ticked them if no problem.

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [ ] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [x] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [x] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [x] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other styles
- [x] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [x] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [x] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [x] 本拉取/合并请求代码是高质量的 Code in this PR is of high quality
- [x] 本拉取/合并使用[formatting](https://github.com/mysterywolf/formatting)等源码格式化工具确保格式符合[RT-Thread代码规范](../documentation/coding_style_cn.md) This PR complies with [RT-Thread code specification](../documentation/coding_style_en.txt)

### CODE REVIEW

File: components/libc/posix/ipc/SConscript [hunk 1]
```diff
@@ -0,0 +1,12 @@
+from building import *
+from utils import VersionCmp
+
+cwd        = GetCurrentDir()
+src        = Glob('*.c')
+CPPPATH    = [cwd]
+CPPDEFINES = []
```
Comment (mysterywolf) [inline components/libc/posix/ipc/SConscript:7]: 删掉吧

**Review by mysterywolf — CHANGES_REQUESTED** [review]

File: components/libc/posix/src/unistd.c [hunk 2]
```diff
@@ -39,3 +39,17 @@ char *ttyname(int fd)
...
+    length = rt_strlen(name);
+
+    if(buf != RT_NULL)
+    {
+        length = rt_strncpy(buf, name, len);
+    }
+
+    return length;
+}
+RTM_EXPORT(confstr);
```
Comment (mysterywolf) [inline components/libc/posix/src/unistd.c:55]: 删掉

**Review by mysterywolf — CHANGES_REQUESTED** [review]

File: components/libc/posix/ipc/SConscript (hunk 1 above)
Comment (xiangxistu) [reply components/libc/posix/ipc/SConscript:7]: image

File: components/libc/compilers/common/nogcc/sys/unistd.h [hunk 3]
```diff
@@ -32,6 +32,7 @@ int access(const char *path, int amode);
 int pipe(int fildes[2]);
 int isatty(int fd);
 char *ttyname(int desc);
+size_t confstr(int name, char *buf, size_t len);
```
[the remark this answers is not in the document]
Comment (xiangxistu) [reply components/libc/compilers/common/nogcc/sys/unistd.h:35]: OK，已处理；

File: components/libc/compilers/common/sys/utsname.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+
+int uname(struct utsname *buf);
+
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif
```
[the remark this answers is not in the document]
Comment (xiangxistu) [reply components/libc/compilers/common/sys/utsname.h:41]: 已处理

File: components/libc/compilers/gcc/newlib/machine/unistd.h [hunk 5]
```diff
@@ -0,0 +1,7 @@
+#ifndef _MACHTIME_H_
+#define _MACHTIME_H_
+
+#include <string.h>
+size_t confstr(int name, char *buf, size_t len);
+
+#endif  /* _MACHTIME_H_ */
```
[the remark this answers is not in the document]
Comment (xiangxistu) [reply components/libc/compilers/gcc/newlib/machine/unistd.h:7]: 已处理

File: components/libc/compilers/common/utsname.c [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+static struct utsname default_utsname;
+
+int uname(struct utsname *buf)
+{
+    rt_memcpy(&default_utsname, buf, sizeof(struct utsname));
+
+    return 0;
+}
+
```
[the remark this answers is not in the document]
Comment (xiangxistu) [reply components/libc/compilers/common/utsname.c:21]: 已处理

File: components/libc/compilers/common/time.c [hunk 7]
```diff
@@ -612,6 +612,17 @@ int clock_gettime(clockid_t clockid, struct timespec *tp)
 }
 RTM_EXPORT(clock_gettime);
 
+int clock_nanosleep(clockid_t clockid, int flags, const struct timespec *rqtp, struct timespec *rmtp)
+{
+    if ((clockid != CLOCK_REALTIME) || (rqtp == RT_NULL))
```
Comment (mysterywolf) [inline components/libc/compilers/common/time.c:617]: sys/time.h没有给出这个函数的声明

File: components/libc/compilers/common/time.c [hunk 8]
```diff
@@ -655,6 +666,33 @@ int clock_settime(clockid_t clockid, const struct timespec *tp)
 }
 RTM_EXPORT(clock_settime);
 
+int nanosleep(const struct timespec *rqtp, struct timespec *rmtp)
+{
+    uint32_t time_ms = rqtp->tv_sec * 1000;
+    uint32_t time_us = rqtp->tv_nsec / 1000;
```
Comment (mysterywolf) [inline components/libc/compilers/common/time.c:672]: sys/time.h没有给出这个函数的声明

**Review by mysterywolf — CHANGES_REQUESTED** [review]

File: components/libc/posix/pthreads/SConscript [hunk 9]
```diff
@@ -11,12 +11,14 @@ libc_ver   = GetDepend('LIBC_VERSION')
 try:
     ver = libc_ver.split(' ')
     ver = ver[1]
```
Comment (mysterywolf) [inline components/libc/posix/pthreads/SConscript:13]: 这个位置怎么加回来了？

**Review by mysterywolf — CHANGES_REQUESTED** [review]

File: components/libc/compilers/common/nogcc/sys/_timespec.h [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+#ifndef _SYS__TIMESPEC_H_
+#define _SYS__TIMESPEC_H_
+
+#include <sys/_types.h>
+
+#if !defined(__time_t_defined) && !defined(_TIME_T_DECLARED)
+typedef _TIME_T_ time_t;
+#define __time_t_defined
+#define _TIME_T_DECLARED
```
Comment (mysterywolf) [inline components/libc/compilers/common/nogcc/sys/_timespec.h:19]: 这个文件删掉

File: components/libc/compilers/gcc/partial/mips/SConscript [hunk 11]
```diff
@@ -11,7 +11,9 @@ group = []
 if rtconfig.PLATFORM == 'gcc' and ('mips' in rtconfig.PREFIX): # identify mips gcc tool chain
     try:
         # There is no 'sys/select.h' in tthe mips gcc toolchain; it will be copied from 'nogcc/sys/select.h'
+        # There is no 'sys/_timespec.h' in tthe mips gcc toolchain; it will be copied from 'nogcc/sys/_timespec.h'
         copy("../../../common/nogcc/sys/select.h", "./sys/select.h")
+        copy("../../../common/nogcc/sys/_timespec.h", "./sys/_timespec.h")
```
Comment (mysterywolf) [inline components/libc/compilers/gcc/partial/mips/SConscript:16]: 删除掉此句

File: components/libc/compilers/common/sys/time.h [hunk 12]
```diff
@@ -53,16 +52,20 @@ struct timeval
 };
 #endif
 #endif /* _TIMEVAL_DEFINED */
+#include <time.h>
 
 #if !(defined(__GNUC__) && !defined(__ARMCC_VERSION)/*GCC*/) && \
     !(defined(__ICCARM__) && (__VER__ >= 8010001)) && \
     !defined(_WIN32)
+#ifndef _TIMESPEC_DEFINED
+#define _TIMESPEC_DEFINED
```
Comment (mysterywolf) [inline components/libc/compilers/common/sys/time.h:61]: revert此文件的修改

File: components/libc/compilers/gcc/newlib/machine/time.h [hunk 13]
```diff
@@ -4,9 +4,7 @@
 #include <rtconfig.h>
 #define _CLOCKS_PER_SEC_   RT_TICK_PER_SECOND
 
-#ifdef __SPU__
 #include <sys/_timespec.h>
 int nanosleep (const struct timespec *, struct timespec *);
-#endif
```
Comment (mysterywolf) [inline components/libc/compilers/gcc/newlib/machine/time.h:9]: 这段删掉吧

**Review by mysterywolf — CHANGES_REQUESTED** [review]

**Comment by mysterywolf** [comment]
把time相关的先都移除吧 这部分我来处理

**Comment by mysterywolf** [comment]
https://github.com/RT-Thread/rt-thread/pull/5384

**Pull request closed without merging**